app.py

```diff
@@ -1015,6 +1017,50 @@ def get_cheatsheet_page(level):
 
     return render_template("cheatsheet.html", commands=commands, level=level)
 
+@app.route('/certificate/<username>', methods=['GET'])
+def get_certificate_page(username):
+    if not current_user()['username']:
+        return utils.error_page(error=403, ui_message=gettext('unauthorized'))
+    username = username.lower()
+    user = DATABASE.user_by_username(username)
+    if not user:
+        return utils.error_page(error=403, ui_message=gettext('user_inexistent'))
+    progress_data = DATABASE.progress_by_username(username)   
+    if progress_data is None:
+        return utils.error_page(error=404, ui_message=gettext('no_certificate'))
+    achievements = progress_data.get('achieved', None)
+    if achievements is None or 'hedy_certificate' not in achievements:
+        return utils.error_page(error=404, ui_message=gettext('no_certificate'))
+    if 'run_programs' in progress_data:
+        count_programs = progress_data['run_programs']
+    else:
+        count_programs = 0
+    quiz_score = get_highest_quiz_score(username)
+    longest_program = get_longest_program(username)
+    
+    number_achievements = len(achievements)
+    congrats_message = gettext('congrats_message').format(**{'username': username})
+    return render_template("certificate.html", count_programs=count_programs, quiz_score=quiz_score, 
+                            longest_program=longest_program, number_achievements=number_achievements, 
+                            congrats_message=congrats_message)
+
+def get_highest_quiz_score(username):
+    max = 0
+    quizzes = DATABASE.get_quiz_stats([username])
+    for quiz in quizzes:
+        if 'scores' in quiz:
+            for score in quiz['scores']:
```

Comment:
Just a minor note on this piece of code, it can be simplified by using `.get()`. Writing it like this would also work:
```
for score in quiz.get('scores', []):
```

Reply:
Good catch, Tibiba! Thank you for pointing it out, will change it right away!!

Reply:
Done! Much cleaner this way!